File: erubit.platform/src/main/java/u/U.java
package u;

import android.content.Context;
import android.os.Build;
import android.text.Html;
import android.text.Spanned;
import android.view.View;
import android.view.animation.Animation;
import android.view.animation.TranslateAnimation;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.util.Arrays;
import java.util.Random;

import a.erubit.platform.android.App;

public class U {
    public static String loadStringResource(Context context, int id) throws IOException {
        Writer writer = new StringWriter();
        char[] buffer = new char[1024];
        try (InputStream is = context.getResources().openRawResource(id)) {
            Reader reader = new BufferedReader(new InputStreamReader(is, "UTF-8"));
            int n;
            while ((n = reader.read(buffer)) != -1) {
                writer.write(buffer, 0, n);
            }
        }

        return writer.toString();
    }

    public static void shuffleIntArray(int[] array)
    {
        int index;
        Random random = new Random();
        for (int i = array.length - 1; i > 0; i--)
        {
            index = random.nextInt(i + 1);
            if (index != i)
            {
                array[index] ^= array[i];
                array[i] ^= array[index];
                array[index] ^= array[i];
            }
        }
    }

    public static void shuffleStrArray(String[] arr) {
        Random rgen = new Random();

        for (int i = 0; i < arr.length; i++) {
            int randPos = rgen.nextInt(arr.length);
            String tmp = arr[i];
            arr[i] = arr[randPos];
            arr[randPos] = tmp;
        }
    }

    public static void animateNegative(View view) {
        Animation animation = new TranslateAnimation(-16, +16, 0, 0);
        animation.setDuration(100);
        animation.setFillAfter(false);
        animation.setRepeatCount(3);
        animation.setRepeatMode(Animation.REVERSE);

        view.startAnimation(animation);
    }

    @SuppressWarnings("deprecation")
    public static Spanned getSpanned(String text) {
        Spanned spanned;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            spanned = Html.fromHtml(text, Html.FROM_HTML_MODE_LEGACY);
        else
            spanned = Html.fromHtml(text);
        return spanned;
    }

    public static String defurigana(String text) {
        // Don't trust Lint, regex is correct
        return text.replaceAll("\\{(.*?):(.*?)\\}", "$1");
    }

    public static boolean equals(String s1, String s2) {
        if (s1 == null || s2 == null)
            return false;

        String d1 = defurigana(s1).replace('　', ' ');
        String d2 = defurigana(s2).replace('　', ' ');

        return d1.equals(d2);
    }

    public static boolean equalsIndependent(String s1, String s2) {
        String regex = "\\s+:\\s+";
        String[] w1 = defurigana(s1).split(regex);
        String[] w2 = defurigana(s2).split(regex);
        Arrays.sort(w1);
        Arrays.sort(w2);

        return Arrays.equals(w1, w2);
    }

    public static String getStringValue(Context context, JSONObject jo, String property) throws JSONException {
        String value = "";
        if (jo.has(property)) {
            Object djo = jo.get(property);
            if (djo instanceof JSONArray) {
                JSONArray jd = jo.getJSONArray(property);
                for (int k = 0; k < jd.length(); k++)
                    value += jd.getString(k);
            } else {
                value = (String)djo;
                if (value.startsWith(C.RESREF)) {
                    String resourceName = value.substring(C.RESREF.length());
                    int resourceId = context.getResources().getIdentifier(resourceName, "raw", context.getPackageName());
                    try {
                        value = U.loadStringResource(context, resourceId);
                    } catch (IOException e) {
                        value = "";
                    }
                }
            }
        }
        return value;
    }
}

File: erubit.platform/src/main/java/u/U.kt
package u

import android.content.Context
import android.os.Build
import android.text.Html
import android.text.Spanned
import android.view.View
import android.view.animation.Animation
import android.view.animation.TranslateAnimation

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.*
import java.util.*


object U {
	@Throws(IOException::class)
	fun loadStringResource(context: Context, id: Int): String {
		val writer: Writer = StringWriter()
		val buffer = CharArray(1024)

		context.resources.openRawResource(id).use { s ->
			val reader: Reader = BufferedReader(InputStreamReader(s, "UTF-8"))
			var n: Int
			while (reader.read(buffer).also { n = it } != -1)
				writer.write(buffer, 0, n)
		}

		return writer.toString()
	}

	fun shuffleIntArray(array: IntArray) {
		var index: Int
		val random = Random()

		for (i in array.size - 1 downTo 1) {
			index = random.nextInt(i + 1)
			if (index != i) {
				array[index] = array[index] xor array[i]
				array[i] = array[i] xor array[index]
				array[index] = array[index] xor array[i]
			}
		}
	}

	fun shuffleStrArray(arr: Array<String?>) {
		val rgen = Random()

		for (i in arr.indices) {
			val randPos = rgen.nextInt(arr.size)
			val tmp = arr[i]
			arr[i] = arr[randPos]
			arr[randPos] = tmp
		}
	}

	fun animateNegative(view: View) {
		val animation: Animation = TranslateAnimation((-16).toFloat(), (+16).toFloat(), 0F, 0F)

		animation.duration = 100
		animation.fillAfter = false
		animation.repeatCount = 3
		animation.repeatMode = Animation.REVERSE

		view.startAnimation(animation)
	}

	fun getSpanned(text: String): Spanned {
		return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
			Html.fromHtml(text, Html.FROM_HTML_MODE_LEGACY)
		else
			Html.fromHtml(text)
	}

	@JvmStatic
	fun defurigana(text: String): String {
		// Don't trust Lint here, regex is correct
		return text.replace("\\{(.*?):(.*?)\\}".toRegex(), "$1")
	}

	fun equals(s1: String?, s2: String?): Boolean {
		if (s1 == null || s2 == null)
			return false

		val d1 = defurigana(s1).replace('　', ' ')
		val d2 = defurigana(s2).replace('　', ' ')

		return d1 == d2
	}

	fun equalsIndependent(s1: String, s2: String): Boolean {
		val regex = "\\s+:\\s+"

		val w1 = defurigana(s1).split(regex).toTypedArray()
		val w2 = defurigana(s2).split(regex).toTypedArray()

		Arrays.sort(w1)
		Arrays.sort(w2)

		return Arrays.equals(w1, w2)
	}

	@Throws(JSONException::class)
	fun getStringValue(context: Context, jo: JSONObject, property: String?): String {
		var value = ""

		if (jo.has(property)) {
			val djo = jo[property]

			if (djo is JSONArray) {
				val jd = jo.getJSONArray(property)
				for (k in 0 until jd.length()) value += jd.getString(k)
			} else {
				value = djo as String
				if (value.startsWith(C.RESREF)) {
					val resourceName = value.substring(C.RESREF.length)
					val resourceId = context.resources.getIdentifier(resourceName, "raw", context.packageName)
					value = try {
						loadStringResource(context, resourceId)
					} catch (e: IOException) {
						""
					}
				}
			}
		}
		return value
	}
}
